specs/capri/Multi-attach.md

```diff
@@ -0,0 +1,151 @@
+```
+type DockSpec struct {
+	*BaseModel
+	Type string `json:"type,omitempty"`
+	Name string `json:"name,omitempty"`
+	Description string `json:"description,omitempty"`
+	Status string `json:"status,omitempty"`
+	StorageType string `json:"storageType,omitempty"`
+	Endpoint string `json:"endpoint,omitempty"`
+	NodeId string `json:"nodeId,omitempty"`
+	DriverName string `json:"driverName,omitempty"`
+	MultiAttach bool `json:"multiAttach,omitempty"`
+	Metadata map[string]string `json:"metadata,omitempty"`	
+}
+```
+
+2. Add MultiAttach field in pool spec
+
+```
+type StoragePoolSpec struct {
+	*BaseModel
+	Name string `json:"name,omitempty"`
+	Description string `json:"description,omitempty"`
+	Status string `json:"status,omitempty"`
+	DockId string `json:"dockId,omitempty"`
+	AvailabilityZone string `json:"availabilityZone,omitempty"`
+	TotalCapacity int64 `json:"totalCapacity,omitempty"`
+	FreeCapacity int64 `json:"freeCapacity,omitempty"`
+	StorageType string `json:"storageType,omitempty"`
+	Extras StoragePoolExtraSpec `json:"extras,omitempty"`
+	ReplicationType string `json:"replicationType,omitempty"`
+	ReplicationDriverName string `json:"replicationDriverName,omitempty"`
+
+	// MultiAttach
+	// If true, this volume can attach to more than one instance. Default will be multiAttach:False
+	MultiAttach bool `json:"multiAttach,omitempty"`
+}
+```
+
+
+3. The volume attachment struct adds AttachMode, read-only (‘ro’) or read-and-write (‘rw’), default is ‘rw’.
+
+```
+type VolumeAttachmentSpec struct {
+	*BaseModel
+	TenantId string `json:"tenantId,omitempty"`
+	UserId string `json:"userId,omitempty"`
+	VolumeId string `json:"volumeId,omitempty"`
+	Mountpoint string `json:"mountpoint,omitempty"`
+	Status string `json:"status,omitempty"`
+	Metadata map[string]string `json:"metadata,omitempty"`
+	HostInfo `json:"hostInfo,omitempty"`
+	ConnectionInfo `json:"connectionInfo,omitempty"`
+	AccessProtocol string `json:"accessProtocol,omitempty"`
+	AttachMode string `json:"attachMode,omitempty"`
+}
+```
+
+4. The volume struct adds MultiAttach.
+
+```
+type VolumeSpec struct {
+	*BaseModel
+	TenantId string `json:"tenantId,omitempty"`
+	UserId string `json:"userId,omitempty"`
+	Name string `json:"name,omitempty"`
+	Description string `json:"description,omitempty"`
+	GroupId string `json:"groupId,omitempty"`
+	Size int64 `json:"size,omitempty"`
+	AvailabilityZone string `json:"availabilityZone,omitempty"`
+	Status string `json:"status,omitempty"`
+	PoolId string `json:"poolId,omitempty"`
+	ProfileId string `json:"profileId,omitempty"`
+	Metadata map[string]string `json:"metadata,omitempty"`
+	SnapshotId string `json:"snapshotId,omitempty"`
+	SnapshotFromCloud bool `json:"snapshotFromCloud,omitempty"`
+	ReplicationId string `json:"replicationId,omitempty"`
+	ReplicationDriverData map[string]string `json:"replicationDriverData,omitempty"`
+	AttachStatus string
+	MultiAttach bool `json:"multiAttach"`
+}
+```
+
+### REST API impact
+
+N/A
```

Comment:
This should affect profiles too.  How does user request a volume with multi-attach capability?

Reply:
Please see performance impact for details.
User first needs to use the following command to create a multi-attach related profile and specify it in volume creation.

```
osdsctl profile create '{"name": "multi-attach", "description": "multi attach profile","customProperties":{"multiAttach":"<is> true"}}'
```

Reply:
volume has the `multiattach` attribute, user can specify it in volume creation body. profile don't have to specify the capability.